Validate that getMenuProps has set the ref correctly

alexandernanberg opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**:
Fixes #490 by using the same approach as we do in `getRootProps`.
<!-- Why are these changes necessary? -->

**Why**:
Because it could lead to unexpected behavior and to avoid confusion.

<!-- How were these changes implemented? -->

**How**:
I'm still unsure of what the best solution for this is. I would like to make the check in the `render` method as we do for the `getRootProps`, but I don't think that's possible because the ref hasn't set `this._menuNode` on the first render. I guess we could loop through all children but that seems unnecessary heavy/slow.

So instead I make the validation/check in `componentDidUpdate`, which looks to work fine.

Another problem I encountered was that there is no way to check if `this._menuNode` is a composite element or not, which makes it hard to use the same validation as we do for `getRootProps`.

I wanted you opinion on this before I continue and add tests and update the documentation. Is there another better way that I've not thought of?

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation
- [x] Tests
- [x] Ready to be merged
      <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [ ] Added myself to contributors table N/A
      <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->

kentcdodds commented on src/downshift.js:
```diff
@@ -983,6 +990,10 @@ class Downshift extends Component {
   }
 
   componentDidUpdate(prevProps, prevState) {
```
Should we add this to a `componentDidMount` as well?

alexandernanberg replied:
Yep that would make sense!

alexandernanberg commented:
Okay, I'll do that! Hopefully I'll have time before I go on vacation on Wednesday. 

That would be a great idea, what do you think of using something like https://www.npmjs.com/package/warning? But I'll probably do this in another PR in that case

kentcdodds commented:
Sounds perfect :+1:

alexandernanberg commented:
Added tests and updated the `getMenuProps` section in the `README.md` now. Basically just copied the tests for `getRootProps` and changed them where it was needed

kentcdodds commented on src/downshift.js:
```diff
@@ -983,6 +994,10 @@ class Downshift extends Component {
   }
 
   componentDidUpdate(prevProps, prevState) {
+    if (this.getMenuProps.called && !this.getMenuProps.suppressRefError) {
```
This should probably be:

```javascript
if (
  this.getMenuProps.called &&
  !this.getMenuProps.suppressRefError &&
  preval`module.exports = process.env.BUILD_REACT_NATIVE !== 'true'`
) {
```

so this code can be eliminated for native. Same with the above.

alexandernanberg replied:
I’ll fix that tomorrow! Hmm any reason why we dont do that for ’getRootProps’ as well? Im not so familiar with RN but why wouldnt we want to validate the ref?

kentcdodds replied:
Yeah, I just checked and the only place we're using `_rootNode` is for codepaths that react native doesn't need. So we could probably do the same thing for those.

I'm thinking about making [a custom macro](https://github.com/kentcdodds/babel-plugin-macros) that will shorten things for us/make these kinds of branches easier to maintain. But that'll be for later :)

alexandernanberg replied:
Oh okay! I've pushed the latest changes now, and I think you should be able to use `--autosquash` to squash them together to one commit

I like the sound of that idea!

alexandernanberg commented:
Hmm this is super weird... the error says that preval is not defined. Tried adding istanbul ignore comments locally but still get the same issue also🤔

kentcdodds commented:
:-( I'm afraid I don't have time to look into this at the moment :-/

alexandernanberg commented:
No worries, I’m off a couple of days but will pick this up again when I get back!

alexandernanberg commented:
Looks like everything is working now, don't really know why `/* istanbul ignore if */` didn't work but `/* istanbul ignore next */` just before preval did 🤔
